metrics-operator/controllers/common/analysis/analysis_evaluator_test.go

```diff
@@ -0,0 +1,316 @@
+				MaximumScore: 20.0,
+				Pass:         true,
+				Warning:      false,
+				ObjectiveResults: []types.ObjectiveResult{
+					{
+						Result: types.TargetResult{},
+						Value:  5.0,
+						Score:  10.0,
+						Error:  nil,
+					},
+					{
+						Result: types.TargetResult{},
+						Value:  5.0,
+						Score:  10.0,
+						Error:  nil,
+					},
+				},
+			},
+			mockedEvaluator: &fake.IObjectiveEvaluatorMock{
+				EvaluateFunc: func(values map[string]string, objective v1alpha3.Objective) types.ObjectiveResult {
+					return types.ObjectiveResult{
+						Result: types.TargetResult{},
+						Value:  5.0,
+						Score:  10.0,
+						Error:  nil,
+					}
+				},
+			},
+		},
+		{
+			name:   "warning scenario",
+			values: map[string]string{},
+			a: v1alpha3.AnalysisDefinition{
+				Spec: v1alpha3.AnalysisDefinitionSpec{
+					Objectives: []v1alpha3.Objective{
+						{
+							Weight: 10,
+						},
+					},
+					TotalScore: v1alpha3.TotalScore{
+						PassPercentage:    80,
+						WarningPercentage: 50,
+					},
+				},
+			},
+			want: types.AnalysisResult{
+				TotalScore:   5.0,
+				MaximumScore: 10.0,
+				Pass:         false,
+				Warning:      true,
+				ObjectiveResults: []types.ObjectiveResult{
+					{
+						Result: types.TargetResult{},
+						Value:  5.0,
+						Score:  5.0,
+						Error:  nil,
+					},
+				},
+			},
+			mockedEvaluator: &fake.IObjectiveEvaluatorMock{
+				EvaluateFunc: func(values map[string]string, objective v1alpha3.Objective) types.ObjectiveResult {
+					return types.ObjectiveResult{
+						Result: types.TargetResult{},
+						Value:  5.0,
+						Score:  5.0,
+						Error:  nil,
+					}
+				},
+			},
+		},
+		{
+			name:   "fail scenario",
+			values: map[string]string{},
+			a: v1alpha3.AnalysisDefinition{
+				Spec: v1alpha3.AnalysisDefinitionSpec{
+					Objectives: []v1alpha3.Objective{
+						{
+							Weight: 10,
+						},
+					},
+					TotalScore: v1alpha3.TotalScore{
+						PassPercentage:    80,
+						WarningPercentage: 50,
+					},
+				},
+			},
+			want: types.AnalysisResult{
+				TotalScore:   0.0,
+				MaximumScore: 10.0,
+				Pass:         false,
+				Warning:      false,
+				ObjectiveResults: []types.ObjectiveResult{
+					{
+						Result: types.TargetResult{},
+						Value:  5.0,
+						Score:  0.0,
+						Error:  nil,
+					},
+				},
+			},
+			mockedEvaluator: &fake.IObjectiveEvaluatorMock{
+				EvaluateFunc: func(values map[string]string, objective v1alpha3.Objective) types.ObjectiveResult {
+					return types.ObjectiveResult{
+						Result: types.TargetResult{},
+						Value:  5.0,
+						Score:  0.0,
+						Error:  nil,
+					}
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ae := NewAnalysisEvaluator(tt.mockedEvaluator)
+			require.Equal(t, tt.want, ae.Evaluate(tt.values, tt.a))
+		})
+	}
+}
+
+// The scenario where the score is passing, but keyObjective fails
+// we cannot mock, therefore we need to use the full slo scoring
```

Comment:
why can this scenario not be mocked?

Reply:
I do not know about the possibility how to mock a component function to return different results on different calls.Basically we are iterating through the objectives and evaluating them. To test this scenario, I would need the first objective (with a big weight) to pass and the second one (with a small weight, but keyObjective) to fail. Then I have a score, that should pass, but I have a objective marked as keyObjective that failed -> the whole analysis should fail (even if according to the score it should have passed)

Reply:
my suggestion would be that for that scenario the mock function could check if the `objective.keyObjective` is set and return `fail`, and return the result with the passing score for the other

Reply:
ah, that might work, thanks will have a look :)